sim/midas/src/test/scala/midas/testutils/FirrtlSpec.scala

```diff
@@ -28,30 +31,46 @@ class CheckLowForm extends SeqTransform {
   )
 }
 
-trait FirrtlRunners extends BackendCompilationUtilities {
+case class RenameTopAnnotation(newTopName: String) extends NoTargetAnnotation
 
-  val cppHarnessResourceName: String = "/firrtl/testTop.cpp"
-  /** Extra transforms to run by default */
-  val extraCheckTransforms = Seq(new CheckLowForm)
+object RenameTop extends Transform with PreservesAll[Transform] {
+  def inputForm = UnknownForm
+  def outputForm = UnknownForm
 
-  private class RenameTop(newTopPrefix: String) extends Transform {
-    def inputForm: LowForm.type = LowForm
-    def outputForm: LowForm.type = LowForm
+  override val optionalPrerequisites = Seq(Dependency[RenameModules])
 
-    def execute(state: CircuitState): CircuitState = {
-      val namespace = state.annotations.collectFirst {
-        case m: ModuleNamespaceAnnotation => m
-      }.get.namespace
+  override val optionalPrerequisiteOf = Seq(Dependency[VerilogEmitter], Dependency[MinimumVerilogEmitter])
 
-      val newTopName = namespace.newName(newTopPrefix)
-      val modulesx = state.circuit.modules.map {
-        case mod: Module if mod.name == state.circuit.main => mod.mapString(_ => newTopName)
-        case other => other
-      }
+  def execute(state: CircuitState): CircuitState = {
+    val c = state.circuit
+    val ns = Namespace(c)
+
+    val newTopName = state.annotations.collectFirst({
+      case RenameTopAnnotation(name) =>
+        require(ns.tryName(name))
+        name
+    }).getOrElse(c.main)
 
-      state.copy(circuit = state.circuit.copy(main = newTopName, modules = modulesx))
+    state.annotations.collect {
+      case ModuleNamespaceAnnotation(mustNotCollideNS) => require(mustNotCollideNS.tryName(newTopName))
     }
+
+    val modulesx = c.modules.map {
+      case m: Module if (m.name == c.main) => m.copy(name = newTopName)
+      case m => m
+    }
+
+    val renames = RenameMap()
+    renames.record(CircuitTarget(c.main), CircuitTarget(newTopName))
+    state.copy(circuit = c.copy(main = newTopName, modules = modulesx), renames = Some(renames))
   }
+}
+
+trait FirrtlRunners extends BackendCompilationUtilities {
```

Comment:
Just curious, what's the blocking thing that makes it necessary to duplicate here?